File: ios/DJMediWallet/DJMediWallet/Models/FHIR/Observation.swift
//
//  Observation.swift
//  DJMediWallet
//
//  FHIR Observation resource model
//

import Foundation

/// FHIR Observation resource for vital signs and lab results
public struct Observation: Codable {
    public let resourceType: String
    public let id: String?
    public let status: String
    public let category: [CodeableConcept]?
    public let code: CodeableConcept
    public let subject: Reference?
    public let effectiveDateTime: String?
    public let issued: String?
    public let valueQuantity: Quantity?
    public let valueString: String?
    public let valueBoolean: Bool?
    public let component: [ObservationComponent]?
    public let interpretation: [CodeableConcept]?
    public let note: [Annotation]?
    
    public init(
        id: String? = nil,
        status: String,
        category: [CodeableConcept]? = nil,
        code: CodeableConcept,
        subject: Reference? = nil,
        effectiveDateTime: String? = nil,
        issued: String? = nil,
        valueQuantity: Quantity? = nil,
        valueString: String? = nil,
        valueBoolean: Bool? = nil,
        component: [ObservationComponent]? = nil,
        interpretation: [CodeableConcept]? = nil,
        note: [Annotation]? = nil
    ) {
        self.resourceType = "Observation"
        self.id = id
        self.status = status
        self.category = category
        self.code = code
        self.subject = subject
        self.effectiveDateTime = effectiveDateTime
        self.issued = issued
        self.valueQuantity = valueQuantity
        self.valueString = valueString
        self.valueBoolean = valueBoolean
        self.component = component
        self.interpretation = interpretation
        self.note = note
    }
}

/// FHIR Observation Component (for multi-value observations)
public struct ObservationComponent: Codable {
    public let code: CodeableConcept
    public let valueQuantity: Quantity?
    public let valueString: String?
    public let valueBoolean: Bool?
    
    public init(
        code: CodeableConcept,
        valueQuantity: Quantity? = nil,
        valueString: String? = nil,
        valueBoolean: Bool? = nil
    ) {
        self.code = code
        self.valueQuantity = valueQuantity
        self.valueString = valueString
        self.valueBoolean = valueBoolean
    }
}

/// FHIR Quantity
public struct Quantity: Codable {
    public let value: Double?
    public let unit: String?
    public let system: String?
    public let code: String?
    
    public init(value: Double?, unit: String? = nil, system: String? = nil, code: String? = nil) {
        self.value = value
        self.unit = unit
        self.system = system
        self.code = code
    }
    
    public var displayValue: String {
        guard let value = value else { return "" }
        if let unit = unit {
            return "\(value) \(unit)"
        }
        return "\(value)"
    }
}

/// FHIR Reference
public struct Reference: Codable {
    public let reference: String?
    public let display: String?
    public let type: String?
    
    public init(reference: String?, display: String? = nil, type: String? = nil) {
        self.reference = reference
        self.display = display
        self.type = type
    }
}

/// FHIR Annotation
public struct Annotation: Codable {
    public let text: String
    public let authorString: String?
    public let time: String?
    
    public init(text: String, authorString: String? = nil, time: String? = nil) {
        self.text = text
        self.authorString = authorString
        self.time = time
    }
}

File: android/DJMediWallet/app/src/main/java/com/djmediwallet/models/fhir/Observation.kt
package com.djmediwallet.models.fhir

/**
 * FHIR Observation resource for vital signs and lab results
 */
data class Observation(
    val resourceType: String = "Observation",
    val id: String? = null,
    val status: String,
    val category: List<CodeableConcept>? = null,
    val code: CodeableConcept,
    val subject: Reference? = null,
    val effectiveDateTime: String? = null,
    val issued: String? = null,
    val valueQuantity: Quantity? = null,
    val valueString: String? = null,
    val valueBoolean: Boolean? = null,
    val component: List<ObservationComponent>? = null,
    val interpretation: List<CodeableConcept>? = null,
    val note: List<Annotation>? = null
)

/**
 * FHIR Observation Component
 */
data class ObservationComponent(
    val code: CodeableConcept,
    val valueQuantity: Quantity? = null,
    val valueString: String? = null,
    val valueBoolean: Boolean? = null
)

/**
 * FHIR Quantity
 */
data class Quantity(
    val value: Double? = null,
    val unit: String? = null,
    val system: String? = null,
    val code: String? = null
) {
    val displayValue: String
        get() = value?.let { v ->
            unit?.let { u -> "$v $u" } ?: v.toString()
        } ?: ""
}

/**
 * FHIR Reference
 */
data class Reference(
    val reference: String? = null,
    val display: String? = null,
    val type: String? = null
)

/**
 * FHIR Annotation
 */
data class Annotation(
    val text: String,
    val authorString: String? = null,
    val time: String? = null
)

/**
 * FHIR Condition resource for diagnoses
 */
data class Condition(
    val resourceType: String = "Condition",
    val id: String? = null,
    val clinicalStatus: CodeableConcept? = null,
    val verificationStatus: CodeableConcept? = null,
    val category: List<CodeableConcept>? = null,
    val severity: CodeableConcept? = null,
    val code: CodeableConcept,
    val subject: Reference,
    val onsetDateTime: String? = null,
    val recordedDate: String? = null,
    val note: List<Annotation>? = null
)

/**
 * FHIR MedicationStatement resource
 */
data class MedicationStatement(
    val resourceType: String = "MedicationStatement",
    val id: String? = null,
    val status: String,
    val medicationCodeableConcept: CodeableConcept? = null,
    val medicationReference: Reference? = null,
    val subject: Reference,
    val effectiveDateTime: String? = null,
    val dateAsserted: String? = null,
    val dosage: List<Dosage>? = null,
    val note: List<Annotation>? = null
)

/**
 * FHIR Dosage
 */
data class Dosage(
    val text: String? = null,
    val timing: Timing? = null,
    val route: CodeableConcept? = null,
    val doseAndRate: List<DoseAndRate>? = null
)

/**
 * FHIR Timing
 */
data class Timing(
    val repeat: TimingRepeat? = null,
    val code: CodeableConcept? = null
)

/**
 * FHIR TimingRepeat
 */
data class TimingRepeat(
    val frequency: Int? = null,
    val period: Double? = null,
    val periodUnit: String? = null
)

/**
 * FHIR DoseAndRate
 */
data class DoseAndRate(
    val type: CodeableConcept? = null,
    val doseQuantity: Quantity? = null
)
